gstaichi/python/dlpack_funcs.cpp

```diff
@@ -11,90 +11,203 @@
   }
-
-  DataType ndarray_data_type = ndarray->get_element_data_type();
-  uint8_t data_type_code = kDLInt;
-
-  uint8_t element_bits = 0;
-  PrimitiveTypeID type_id = ndarray_data_type->as<PrimitiveType>()->type;
+  PrimitiveTypeID type_id = dt_as_primitive->type;
   switch (type_id) {
     case PrimitiveTypeID::i32: {
-      data_type_code = static_cast<uint8_t>(kDLInt);
-      element_bits = 32;
+      *p_data_type_code = static_cast<uint8_t>(kDLInt);
+      *p_element_bits = 32;
       break;
     }
     case PrimitiveTypeID::i64: {
-      data_type_code = static_cast<uint8_t>(kDLInt);
-      element_bits = 64;
+      *p_data_type_code = static_cast<uint8_t>(kDLInt);
+      *p_element_bits = 64;
       break;
     }
     case PrimitiveTypeID::f32: {
-      data_type_code = static_cast<uint8_t>(kDLFloat);
-      element_bits = 32;
+      *p_data_type_code = static_cast<uint8_t>(kDLFloat);
+      *p_element_bits = 32;
       break;
     }
     case PrimitiveTypeID::f64: {
-      data_type_code = static_cast<uint8_t>(kDLFloat);
-      element_bits = 64;
+      *p_data_type_code = static_cast<uint8_t>(kDLFloat);
+      *p_element_bits = 64;
       break;
     }
     case PrimitiveTypeID::u1: {
-      data_type_code = static_cast<uint8_t>(kDLBool);
-      element_bits = 8;
+      *p_data_type_code = static_cast<uint8_t>(kDLBool);
+      *p_element_bits = 8;
       break;
     }
     default: {
       TI_ERROR("unsupported ndarray data type for dlpack");
     }
   }
+}
+
+pybind11::capsule field_to_dlpack(Program *program,
+                                  SNode *snode,
+                                  int element_ndim,
+                                  int n,
+                                  int m) {
+  if (!snode->is_path_all_dense) {
+    TI_ERROR("Only dense fields are supported for dlpack conversion");
+  }
+
+  Arch arch = program->compile_config().arch;
+  validate_arch(arch);
+
+  int tree_id = snode->get_snode_tree_id();
+  DevicePtr tree_device_ptr = program->get_snode_tree_device_ptr(tree_id);
+
+  int field_in_tree_offset = program->get_field_in_tree_offset(tree_id, snode);
+
+  DLDeviceType device_type = DLDeviceType::kDLCPU;
+  void *raw_ptr = (void *)((uint64_t)get_raw_ptr(arch, program, tree_device_ptr,
+                                                 &device_type) +
+                           field_in_tree_offset);
+
+  DataType dt = snode->dt;
+
+  uint8_t element_bits = 32;
+  uint8_t data_type_code = kDLInt;
+  get_type_info(dt, &data_type_code, &element_bits);
+
+  int ndim = snode->num_active_indices;
+  int full_ndim = ndim + element_ndim;
+  int64_t *shape = nullptr;
+  if (full_ndim > 0) {
+    shape = new int64_t[full_ndim];
+    for (int i = 0; i < ndim; i++) {
+      if (snode->physical_index_position[i] != i) {
+        TI_ERROR(
+            "SNode has non-sequential physical index mapping, which is not "
+            "supported currently for dlpack conversion");
```

Comment:
It would be nice to rather focus on what the user could do the prevent this.

Reply:
Seems this check didn't actually work.
- added test for non-sequential axes
- fixed implementation (AI'd this bit for now)
- modified message slightly